src/Pos/Txp/Core/Types.hs

```diff
@@ -242,25 +262,40 @@ data TxPayload = TxPayload
       -- Basically, address distributions are needed so that (potential)
       -- receivers of P2SH funds would count as stakeholders.
       _txpDistributions :: ![TxDistribution]
-    , -- | Transaction witnesses. Invariant: there are as many witnesses
-      -- as there are transactions in the block. This is checked during
-      -- deserialisation. We can't put them into the same Merkle tree
-      -- with transactions, as the whole point of segwit is to separate
-      -- transactions and witnesses.
-      --
-      -- TODO: should they be put into a separate Merkle tree or left as
-      -- a list?
-      _txpWitnesses     :: ![TxWitness]
     } deriving (Show, Eq, Generic)
 
 makeLenses ''TxPayload
 
+-- | Smart constructor of 'TxPayload' which can fail if invariant is
```

Comment:
Can you add a comment which either
a) says which invariants should be satisfied, or
b) links to those invariants described elsewhere?